WIP openstack: make security groups optional

clnperez opened the pull request:
Some openstack-based offerings, such as PowerVC, do not support security
 groups. Allow users to create a cluster without them. The default is to create
 the security groups (disable = false).
    
 Fixes #4555
  
Signed-off-by: Christy Norman <christy@linux.vnet.ibm.com>

clnperez commented:
/retest

staebler commented on data/data/openstack/main.tf:
```diff
@@ -70,22 +72,23 @@ module "masters" {
+  octavia_support       = var.openstack_octavia_support
+  machines_subnet_id    = var.openstack_machines_subnet_id
+  machines_network_id   = var.openstack_machines_network_id
+  master_extra_sg_ids   = var.openstack_disable_sg ? null : var.openstack_master_extra_sg_ids
```
Do we need to enforce that null is used for the extra security group IDs? With `openstack_disable_sg` being set to true, `master_extra_sg_ids` will not be used in the module anyway.

staebler commented on data/data/openstack/masters/variables.tf:
```diff
@@ -8,6 +8,10 @@ variable "cluster_id" {
   description = "The identifier for the cluster."
 }
 
+variable "openstack_disable_sg" {
```
This variable is unused.

staebler commented on pkg/types/openstack/platform.go:
```diff
@@ -84,4 +84,8 @@ type Platform struct {
+
+	// DisableSecurityGroups will allow clusters to be created without the use of OpenStack Security Groups. Some Openstack
+	// configurations do not support security groups. This is generally not recommended where security groups are supported.
+	DisableSecurityGroups bool `json:"disableSecurityGroups,omitempty"`
```
Is there a way for the installer to determine whether the OpenStack configuration supports security groups without requiring the user to specify this?

clnperez replied:
I wondered about that as well, and am not sure. If someone knows, this would be a good add. Maybe just in a follow-up patch.

staebler replied:
Note that what I am suggesting is that if there is a way for the installer to determine this, then we should not even have it as an option for the user. In other words, we would eliminate this field.

clnperez commented:
/retest

clnperez commented:
Thanks @staebler for the review. I pushed changes addressing your comments to a new commit (will squash these later). The openstack-related jobs here did all pass on the previous run, so, that is encouraging.

staebler commented on data/data/openstack/masters/main.tf:
```diff
@@ -63,7 +63,7 @@ resource "openstack_compute_instance_v2" "master_conf_0" {
   flavor_id = data.openstack_compute_flavor_v2.masters_flavor.id
   image_id = var.root_volume_size == null ? var.base_image_id : null
-  security_groups = var.master_sg_ids
+  security_groups = var.master_sg_ids == null ? [] : var.master_sg_ids
```
This should not be needed. (1) The resource should be able to handle null. (2) The variable will never be null: It is an empty list when security groups are disabled.

clnperez commented:
/test e2e-openstack

clnperez commented:
Closing for now, as priorities have swizzled
